runtime/drivers/clickhouse/embed.go

```diff
@@ -0,0 +1,377 @@
+package clickhouse
+
+import (
+	"archive/tar"
+	"bufio"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/rilldata/rill/cli/pkg/dotrill"
+	"go.uber.org/zap"
+)
+
+const embedVersion = "24.7.4.51"
+
+type embedClickHouse struct {
+	tcpPort int
+	dataDir string
+	tempDir string
+	logger  *zap.Logger
+	cmd     *exec.Cmd
+}
+
+func newEmbedClickHouse(tcpPort int, dataDir, tempDir string, logger *zap.Logger) *embedClickHouse {
+	return &embedClickHouse{tcpPort: tcpPort, dataDir: dataDir, tempDir: tempDir, logger: logger}
+}
+
+// start installs (depending on OS and platform) and starts ClickHouse server.
+// The destination directory for the ClickHouse binary is .rill/clickhouse.
+// The function returns the DSN for the ClickHouse server and close function.
+func (e *embedClickHouse) start() (*clickhouse.Options, error) {
+	// Store the ClickHouse binary under .rill/clickhouse so that every project can use the same binary
+	destDir, err := dotrill.ResolveFilename("clickhouse", true)
+	if err != nil {
+		return nil, err
+	}
+
+	binPath, err := e.install(destDir, e.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if e.tcpPort == 0 {
+		e.tcpPort, err = getFreePort()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	configPath, err := e.prepareConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	e.cmd = exec.Command(binPath, "server", "--config-file", configPath)
+
+	ready := make(chan error, 1)
+	go func() {
+		err := e.startAndWaitUntilReady()
+		ready <- err
+		if err != nil && e.cmd != nil && e.cmd.Process != nil {
+			_ = e.cmd.Process.Kill()
+			return
+		}
+
+		if err := e.cmd.Wait(); err != nil {
+			e.logger.Error("clickhouse server exited with an error", zap.Error(err))
+		}
+	}()
+
+	if err := <-ready; err != nil {
+		return nil, err
+	}
+
+	addr := net.JoinHostPort("localhost", fmt.Sprintf("%d", e.tcpPort))
+	e.logger.Info("ClickHouse server: " + "clickhouse://" + addr)
+
+	return &clickhouse.Options{
+		Protocol: clickhouse.Native,
+		Addr:     []string{addr},
+	}, nil
+}
+
+func (e *embedClickHouse) stop() error {
+	e.logger.Info("Stopping ClickHouse server: " + fmt.Sprintf("localhost:%d", e.tcpPort))
```

Comment:
Should use structured logging and same format as in `start`:

```suggestion
	addr := net.JoinHostPort("localhost", fmt.Sprintf("%d", e.tcpPort))
	e.logger.Info("Stopping embedded ClickHouse server", zap.String("addr", addr))
```